File: graphs/Minimum Obstacle Removal to Reach Corner.py
from collections import deque
class Solution:
    def minimumObstacles(self, grid: List[List[int]]) -> int:
        visited = set((0, 0))
        queue = deque([((0,0), 0)])
        while queue:
            node, length = queue.popleft()
            x, y = node[0], node[1]

            if x == len(grid)-1 and y == len(grid[0])-1:
                return length

            if x > 0 and (x-1, y) not in visited:
                if grid[x-1][y]:
                    queue.append(((x-1, y), length+1))
                else:
                    queue.appendleft(((x-1, y), length))
                visited.add((x-1, y))
            
            if x < len(grid)-1 and (x+1, y) not in visited:
                if grid[x+1][y]:
                    queue.append(((x+1, y), length+1))
                else:
                    queue.appendleft(((x+1, y), length))
                visited.add((x+1, y))
            
            if y > 0 and (x, y-1) not in visited:
                if grid[x][y-1]:
                    queue.append(((x, y-1), length+1))
                else:
                    queue.appendleft(((x, y-1), length))
                visited.add((x, y-1))
            
            if y < len(grid[0])-1 and (x, y+1) not in visited:
                if grid[x][y+1]:
                    queue.append(((x, y+1), length+1))
                else:
                    queue.appendleft(((x, y+1),length))
                visited.add((x, y+1))

File: graphs/Minimum Obstacle Removal to Reach Corner.cpp
struct pair_hash {
    inline std::size_t operator()(const std::pair<int,int> & v) const {
        return v.first*31+v.second;
    }
};


class Solution {
public:
    void dfs(int x, int y, queue<pair<int, int>> &q, unordered_set<pair<int, int>, pair_hash> &seen, vector<vector<int>> &distances, vector<vector<int>> &grid){
        if(x > 0 and seen.find({x-1, y}) == seen.end()){
            seen.insert({x-1, y});
            if(grid[x-1][y]){
                distances[x-1][y] = distances[x][y]+1;
                q.push({x-1, y});
            }else{
                distances[x-1][y] = distances[x][y];
                dfs(x-1, y, q, seen, distances, grid);
            }
        }


        if(x < grid.size()-1 and seen.find({x+1, y}) == seen.end()){
            seen.insert({x+1, y});
            if(grid[x+1][y]){
                distances[x+1][y] = distances[x][y]+1;
                q.push({x+1, y});
            }else{
                distances[x+1][y] = distances[x][y];
                dfs(x+1, y, q, seen, distances, grid);
            }
        }


        if(y > 0 and seen.find({x, y-1}) == seen.end()){
            seen.insert({x, y-1});
            if(grid[x][y-1]){
                distances[x][y-1] = distances[x][y]+1;
                q.push({x, y-1});
            }else{
                distances[x][y-1] = distances[x][y];
                dfs(x, y-1, q, seen, distances, grid);
            }
        }

        if(y < grid[0].size()-1 and seen.find({x, y+1}) == seen.end()){
            seen.insert({x, y+1});
            if(grid[x][y+1]){
                distances[x][y+1] = distances[x][y]+1;
                q.push({x, y+1});
            }else{
                distances[x][y+1] = distances[x][y];
                dfs(x, y+1, q, seen, distances, grid);
            }
        }
    }

    
    int minimumObstacles(vector<vector<int>>& grid) {
        queue<pair<int, int>> q;
        q.push({0, 0});
        unordered_set<pair<int, int>, pair_hash> seen;
        vector<vector<int>> distances(grid.size(), vector<int>(grid[0].size(), 0));
        seen.insert({0, 0});
        while(!q.empty()){
            int x = q.front().first;
            int y = q.front().second;
            q.pop();
            dfs(x, y, q, seen, distances, grid);
        }
        return distances[grid.size()-1][grid[0].size()-1];
    }
};
